modules/apps/document-library/document-library-document-conversion/src/main/java/com/liferay/document/library/document/conversion/internal/DocumentConversionImpl.java

```diff
@@ -133,17 +138,70 @@ else if (!inputDocumentFormat.isExportableTo(outputDocumentFormat)) {
 		UnsyncByteArrayOutputStream unsyncByteArrayOutputStream =
 			new UnsyncByteArrayOutputStream();
 
-		DocumentConverter documentConverter = _getDocumentConverter();
+		long start = System.currentTimeMillis();
 
-		documentConverter.convert(
-			inputStream, inputDocumentFormat, unsyncByteArrayOutputStream,
-			outputDocumentFormat);
+		long timeout = Math.max(
+			PropsValues.DL_FILE_ENTRY_PREVIEW_GENERATION_TIMEOUT_GHOSTSCRIPT,
+			PropsValues.DL_FILE_ENTRY_PREVIEW_GENERATION_TIMEOUT_PDFBOX);
 
-		FileUtil.write(
-			file, unsyncByteArrayOutputStream.unsafeGetByteArray(), 0,
-			unsyncByteArrayOutputStream.size());
+		InputStream finalInputStream = inputStream;
 
-		inputStream.close();
+		Future<?> future = _noticeableExecutorService.submit(
+			() -> {
+				try {
+					DocumentConverter documentConverter =
+						_getDocumentConverter();
+
+					documentConverter.convert(
+						finalInputStream, inputDocumentFormat,
+						unsyncByteArrayOutputStream, outputDocumentFormat);
+
+					FileUtil.write(
+						file, unsyncByteArrayOutputStream.unsafeGetByteArray(),
+						0, unsyncByteArrayOutputStream.size());
+
+					if (_log.isInfoEnabled()) {
+						_log.info(
+							StringBundler.concat(
+								"Conversion from ",
+								inputDocumentFormat.getName(), " to ",
+								outputDocumentFormat.getName(), " in ",
+								System.currentTimeMillis() - start, " ms"));
+					}
+				}
+				catch (IOException ioException) {
+					throw new RuntimeException(ioException);
+				}
+				finally {
+					try {
+						finalInputStream.close();
+					}
+					catch (IOException ioException) {
+						_log.error("Unable to close input stream", ioException);
+					}
+				}
+			});
+
+		try {
+			future.get(timeout, TimeUnit.SECONDS);
+		}
+		catch (TimeoutException timeoutException) {
+			String errorMessage =
+				"Timeout when converting for " + file.getPath();
+
+			if (future.cancel(true)) {
+				errorMessage += " resulted in a canceled timeout for " + future;
+			}
+
+			_log.error(errorMessage);
+
+			throw new IOException(timeoutException);
+		}
+		catch (Exception exception) {
+			_log.error(exception);
+
+			throw new IOException(exception);
+		}
```

Comment:
This looks a bit too complicated. Can you extract the inner part to a private method? Something like this:
```
try {
    _convert(inputStream);
} catch (IOException ...) {
    throw ioException;
} catch (Exception ...) {
   ...
}

...
private void _convert(InputStream inputStream) throws IOException {
    Future<?> future = _noticeableExecutorService.submit(...);
    try {
        int timeout = ...;

        future.get(timeout, ...);
    } catch (TimeoutException ...) {
        ...
    }
}
```
Other than making the main method a bit smaller, it will remove the unnecessary `finalInputStream` variable and move the `timeout` declaration closer to its usage.

Reply:
thanks for the suggestion.